### DISCUSSION TITLE
AI-RBI Hackathon projects

### DISCUSSION THREAD

**Discussion by casperdcl** [SyneRBI/SIRF-Contribs#31] [Ideas]
> [!NOTE]
> All resultant issues/PRs are at [SyneRBI/projects/13](https://github.com/orgs/SyneRBI/projects/13).

Suggestions for the [AI-RBI hackathon](https://www.ccpsynerbi.ac.uk/events/airbi-hackathon/) on 11 & 12 March 2026.

For inspiration, see also issues/PRs in:
- https://github.com/SyneRBI/SIRF
  + https://github.com/SyneRBI/SIRF-Exercises
  + https://github.com/SyneRBI/SIRF-Contribs
  + https://github.com/SyneRBI/SIRF-SuperBuild
- https://github.com/UCL/STIR
- https://github.com/TomographicImaging/CIL
  + https://github.com/TomographicImaging/CIL-Demos

> [!TIP]
> See https://github.com/deepinv/deepinv/discussions/700 for examples of how to write a suggestion in this discussion.

**Comment by KrisThielemans** [comment]
## Project 1: Merge pytorch support framework to SIRF

**Leads**: @casperdcl @KrisThielemans 

**Idea**: See https://github.com/SyneRBI/SIRF/pull/1305. This PR is essentially ready, but throws up some errors in CI. We hope to complete this before the hackathon, but...

We could (possibly trivially) also update it to change `as_array()` calls to `asarray()` to avoid data copies.

**Reply by KrisThielemans** [reply]
Merged!!

**Comment by ckolbPTB** [comment]
## Project 2: Proof-of-concept for MR DL-reconstruction
**Leads:** @ckolbPTB 

**Idea:** Include the SIRF MR acquisition model into a deep inverse DL reconstruction. 

Starting points: 

- MR brainweb SIRF acquisition model: https://github.com/SyneRBI/SIRF-Exercises/blob/master/notebooks/Introductory/acquisition_model_mr_pet_ct.ipynb
- handling of 3D brainweb data in deep inverse: https://deepinv.github.io/deepinv/auto_examples/physics/demo_mri_tour.html#explore-3d-mri
- MODL-based image reconstruction: https://deepinv.github.io/deepinv/auto_examples/self-supervised-learning/demo_artifact2artifact.html#define-model
- Linear physics operator: https://deepinv.github.io/deepinv/api/stubs/deepinv.physics.LinearPhysics.html
- Use a DeepInverse denoiser as proximal step in minimisation https://github.com/TomographicImaging/CIL-User-Showcase/blob/main/016_cil_torch_fista_pnp/fista_with_denoiser.ipynb

Main limitation: No GPU-support for MR acquisition model available yet in SIRF

**Comment by mehrhardt** [comment]
## Project 3: Plug and play regularization for PET
**Leads:** @mehrhardt

**Idea:** Include trained regularizers from DeepInverse into PET reconstruction in SIRF.

- Use pretrained convex ridge regularizer or input convex neural network in either gradient-based or primal-dual PET reconstruction.
- Use pretrained denoiser to replace proximal operator in either gradient-based or primal-dual PET reconstruction.
- Since everything is trained already, it would be easiest to call DeepInverse from SIRF and use SIRF algorithms for reconstruction.

**Comment by mehrhardt** [comment]
## Project 4: Deep Image Prior for PET
**Leads:** @mehrhardt TBC

**Idea:** Use deep image prior for PET reconstruction.

- This will need to have a PET forward model in DeepInverse and could be a good use-case.
- Alternatively, this could also be done with neural fields if this is more interesting to the participants.

**Reply by Andrewwango** [reply]
[Deep Image Prior](https://deepinv.github.io/deepinv/api/stubs/deepinv.models.DeepImagePrior.html#deepinv.models.DeepImagePrior) in DeepInverse

**Comment by ckolbPTB** [comment]
## Project 5: Evaluation of MR DL-reconstruction
**Leads:** @ckolbPTB @paskino 

**Idea:** Evaluate artefacts on commercial DL-reconstruction by comparing it to SIRF/CIL reconstructions. 

**Background:** DL reconstruction provided by vendor leads to artefacts in the reconstructed knee images which are mistaken for pathologies. Patients had to be recalled for a second imaging session. SIRF/CIL reconstruction could help to understand where these artefacts come from and maybe even avoid recalling patients.

Starting point:

- https://github.com/SyneRBI/SIRF-Contribs/blob/recon-comparison-grappa-dl/src/notebooks/Grappa_and_CIL.ipynb

**Comment by Andrewwango** [comment]
# Project 6: DeepInverse wrapper for PET operators

**wrapping e.g. parallelproj/SIRF/CIL/pytomography**

**Leads**: @andrewwango, @tachella

**Idea**: DeepInverse abstracts imaging forward operators into a `physics`, allowing them to be used with all other modern DL tooling in DeepInverse e.g. pretrained models, diffusion sampling, self-supervised learning. We should create a `physics` that wraps SIRF/CIL operators (e.g. PET acquisition), so that users can use this physics natively with DeepInverse algorithms and tools.

Starting point:

The wrapper could follow our Astra integration [`TomographyWithAstra`](https://deepinv.github.io/deepinv/api/stubs/deepinv.physics.TomographyWithAstra.html) and look something like:

```python
class deepinv.physics.EmissionTomographyWithSIRF(deepinv.physics.LinearPhysics):
     ...
```

Tasks:
- Allowing efficient array-to-tensor and device transfer when objects are on CUDA
- Write end-to-end example, like [Low-dose CT with ASTRA backend and Total-Variation (TV) prior](https://deepinv.github.io/deepinv/auto_examples/external-libraries/demo_astra_tomography.html)
- Unit test physics, ensuring adjointness and normalisation

New to DeepInverse? [Tutorial: 5-min quickstart tutorial](https://deepinv.github.io/deepinv/auto_examples/basics/demo_quickstart.html), [Tutorial: how to contribute](https://deepinv.github.io/deepinv/contributing.html)

**Comment by Andrewwango** [comment]
# Project 7: DeepInverse modality-specific utils (generators, IO, vis...) from SIRF

**Leads**: @andrewwango, @tachella

**Idea**: Aside from imaging physics and recon algorithms, DeepInverse users benefit from other modality-specific utils e.g.:
- [Physics generators](https://deepinv.github.io/deepinv/user_guide/physics/intro.html#physics-generators): generate random physics params (PSFs, noise) for simulation
- [IO](https://deepinv.github.io/deepinv/user_guide/other/utils.html#image-loading): expand the current IO library to handle better dicom-to-tensor (and vice versa, or other formats) specifically for PET/SPECT conventions
- [Visualisation](https://deepinv.github.io/deepinv/user_guide/other/utils.html#plotting): support standard 3D viewer tools used in PET/SPECT/MR imaging
- **Document all the above into an end-to-end example**

New to DeepInverse? [Tutorial: 5-min quickstart tutorial](https://deepinv.github.io/deepinv/auto_examples/basics/demo_quickstart.html), [Tutorial: how to contribute](https://deepinv.github.io/deepinv/contributing.html)

**Comment by KrisThielemans** [comment]
## Project 8: update STIR data to CUDA managed pointers

**Leads**: @casperdcl @KrisThielemans @evgueni-ovtchinnikov 

**Idea**: Follow-up from https://github.com/UCL/STIR/discussions/1683#discussioncomment-15935042. 

needed for Project 6 https://github.com/SyneRBI/SIRF-Contribs/discussions/31#discussioncomment-15962068

- likely need to complete the STIR work
- adapt `sirf.STIR` data-containers (images first)
- expose underlying CUDA managed pointer to python

**Reply by KrisThielemans** [reply]
- Merged PRs:
  -  STIR: sort out problems with keeping arrays CUDA-managed:
https://github.com/UCL/STIR/pull/1690, https://github.com/UCL/STIR/pull/1693, https://github.com/UCL/STIR/pull/1694
  - SIRF: https://github.com/SyneRBI/SIRF/pull/1373 fixes Python numerical operations with numpy objects on the "left"
  - parallelproj: small issues with the feedstock and CUDA version pinning
  - https://github.com/UCL/STIR/pull/1689 interfaces to libparallelproj 2.0 and makes "internal" pointers in parallelproj and the Gibbs prior hierarchy CUDA-managed via CuVec
- In-progress PRs:
  - https://github.com/UCL/STIR/pull/1695
  - https://github.com/SyneRBI/SIRF/pull/1381 added CUDA array interface to SIRF (by checking underlying pointer)
 
Other things:
- proof of concept available by @Dimitra-Kyriakopoulou, will be adjusted as we go
- still need to make STIR changes to have `stir::Array` directly (optionally?) use `CuVec`

**Reply by KrisThielemans** [reply]
- https://github.com/UCL/STIR/pull/1695 was merged on 19Mar2026

**Comment by MaxTousss** [comment]
## Project 9: Enable the use of DeepInverse as plug-in in CASToR  

Leads: @sstute1

**Idea**: CASToR enables, among other things, reconstruction of PET list-mode data. By construction, it is ill adapted to be interfaced inside Python. However, we could use python module as a plug-in for image-space processing, thus making it possible to use deep-learning models in key steps of a reconstruction scheme. 

**Tasks**:
- Validate the capability of using python plug-in in CASToR, including pytorch with GPU 
- Implements an algorithm with a pre-learned deep-learning component (e.g., denoiser)
- Implements an algorithm with deep-learning component that is trained in the reconstruction process (e.g., DIP)
